lib/spack/env/cc

```diff
@@ -58,59 +69,140 @@ parameters=(
+    esac
+}
+
+# prepend LISTNAME ELEMENT [SEP]
+#
+# Prepends element to list, assuming list is separted by SEP.
+# Handles empty lists and single-element lists.
+prepend() {
+    varname="$1"
+    elt="$2"
+
+    if empty "$varname"; then
+        eval "$varname=\"\${elt}\""
+    else
+        # get appropriate separator for the list we're appending to
+        setsep "$varname"
+        eval "$varname=\"\${elt}${sep}\${$varname}\""
+    fi
+}
+
+# append LISTNAME ELEMENT [SEP]
+#
+# Appends element to list, assuming list is separted by SEP.
+# Handles empty lists and single-element lists.
+append() {
+    varname="$1"
+    elt="$2"
+
+    if empty "$varname"; then
+        eval "$varname=\"\${elt}\""
+    else
+        # get appropriate separator for the list we're appending to
+        setsep "$varname"
+        eval "$varname=\"\${$varname}${sep}\${elt}\""
+    fi
+}
+
+# extend LISTNAME1 LISTNAME2 [PREFIX]
+#
+# Append elements from second list to first.
+# If PREFIX is provided, prepend it to each element.
+extend() {
+    # figure out appropriate IFS for the list we're reading
+    setsep "$2"
+    if [ "$sep" != " " ]; then
+        IFS="$sep"
+    fi
+    eval "for elt in \${$2}; do append $1 \"$3\${elt}\"; done"
+    unset IFS
+}
+
+# preextend LISTNAME1 LISTNAME2 [PREFIX]
+#
+# Prepend elements from second list to first, preserving order.
+# If PREFIX is provided, prepend it to each element.
+preextend() {
+    # figure out appropriate IFS for the list we're reading
+    setsep "$2"
+    if [ "$sep" != " " ]; then
+        IFS="$sep"
+    fi
+
+    # first, reverse the list to prepend
+    _reversed_list=""
+    eval "for elt in \${$2}; do prepend _reversed_list \"$3\${elt}\"; done"
 
-# SPACK_<LANG>FLAGS and SPACK_LDLIBS are split by ' '
-IFS=' ' read -ra SPACK_FFLAGS   <<< "$SPACK_FFLAGS"
-IFS=' ' read -ra SPACK_CPPFLAGS <<< "$SPACK_CPPFLAGS"
-IFS=' ' read -ra SPACK_CFLAGS   <<< "$SPACK_CFLAGS"
-IFS=' ' read -ra SPACK_CXXFLAGS <<< "$SPACK_CXXFLAGS"
-IFS=' ' read -ra SPACK_LDFLAGS  <<< "$SPACK_LDFLAGS"
-IFS=' ' read -ra SPACK_LDLIBS   <<< "$SPACK_LDLIBS"
+    # prepend reversed list to preextend in order
+    IFS="${lsep}"
+    for elt in $_reversed_list; do prepend $1 "$3${elt}"; done
+    unset IFS
+}
 
+# system_dir PATH
 # test whether a path is a system directory
-function system_dir {
+system_dir() {
+    IFS=':'  # SPACK_SYSTEM_DIRS is colon-separated
     path="$1"
-    for sd in "${SPACK_SYSTEM_DIRS[@]}"; do
-        if [ "${path}" == "${sd}" ] || [ "${path}" == "${sd}/" ]; then
+    for sd in $SPACK_SYSTEM_DIRS; do
+        if [ "${path}" = "${sd}" ] || [ "${path}" = "${sd}/" ]; then
             # success if path starts with a system prefix
+            unset IFS
             return 0
         fi
     done
+    unset IFS
     return 1  # fail if path starts no system prefix
 }
 
-for param in "${parameters[@]}"; do
-    if [[ -z ${!param+x} ]]; then
+for param in $params; do
+    if [ "${param=unset}" = unset ]; then
```

Comment:
We should add a sanity check that the bell character doesn't appear in any arg in `$@`. It won't affect performance much to check, and debugging if that goes wrong will be a nightmare.

Reply:
If someone can actually come up with a case, for some reason my brain wouldn't let me stop thinking about this after dinner until I wrote up this variant that has no separator limitations, because it doesn't need them:

```sh
# <list> <variable> <prefix>
aal_appendv() {
    eval "\
    aal_append_key=\${$1_cnt:=0}
    $1_cnt=\"\$((\${$1_cnt} + 1))\"
    eval \"aaprefix_$1_\$aal_append_key=\\\"$3\\\$\$2\\\"\"
    "
}
# <list> <value> <prefix>
aal_append() {
    aal_append_local="$2"
    aal_appendv "$1" aal_append_local "$3"
}

# extend LISTNAME1 LISTNAME2 [PREFIX]
#
# Append elements from second list to first.
# If PREFIX is provided, prepend it to each element.
aal_extend() {
    aal_extend_cnt=0
    aal_cnt $2 l2_cnt
    while [ $aal_extend_cnt -lt $l2_cnt ] ; do
        # aal_at $2 $aal_extend_cnt aal_extend_val
        aal_appendv "$1" "aaprefix_$2_$aal_extend_cnt" "$3"
        aal_extend_cnt=$(($aal_extend_cnt+1))
    done
}


# <list> <target>
aal_cnt() {
    eval "\
    $2=\${$1_cnt:=0}
    "
}

# <list> <index> <target>
aal_at() {
    eval "\
    $3=\$aaprefix_$1_$2
    "
}

# <list> <target> <join-str>
aal_join() {
    aal_join_str="${3:- }"
    aal_join_cnt=1
    eval "$2=\"\$aaprefix_$1_0\""
    aal_cnt $1 l2_cnt

    while [ $aal_join_cnt -lt $l2_cnt ] ; do
        eval "
            $2=\"\$$2\$aal_join_str\$aaprefix_$1_$aal_join_cnt\"
            "
        aal_join_cnt=$(($aal_join_cnt+1))
    done
}
```

Other possible upside, it seems to be slightly faster at least on a "do a bunch of appends then join/print" microbenchmark.  Bash arrays are still faster by roughly 5x, but still.

The downside is one has to call "join" at the end, since the list isn't actually stored as a string but as a string of numbered entries in the variable dict.  Otherwise it basically has the same interface, no separator issues, etc.  Also I suppose there's the issue that it jams a tremendous number of variables into the global scope. 🤷

Reply:
Let's keep what we've got for now and add an interface like this in a new PR, especially if it is faster.  I do want to do a blog post about these `sh` list interfaces, and this is super cool.

Reply:
I updated this.  Also it turns out this line was wrong -- it doesn't actually check that param is empty -- so I fixed it and added a test.